### PR TITLE
[RFC] Provide optional switches to _dump_nccl_trace

### PR DESCRIPTION
**Pull request by c-p-i-o**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #127042
* __->__ #127651

Summary:
Data from PyTorch distributed is mostly useful during initial stages of model development.
Provide options to reduce data sent/dumped.
`_dump_nccl_trace` takes 3 optional switches. Default as before returns everything
- `includeCollectives`: option to also include collectives: Default is True.
- `includeStacktraces`: option to include stack traces in collectives. Default is True.
- `onlyActive`: option to only send active collective work - i.e. not completed. Default is
    False (i.e. send everything)

Test Plan:
Unit tests

Reviewers:

Subscribers:

Tasks:

Tags:

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

### CODE REVIEW

File: torch/csrc/distributed/c10d/init.cpp [hunk 1]
```diff
@@ -3160,9 +3160,20 @@ such as `dist.all_reduce(tensor, async_op=True)`.
         Arguments:
           tensors(List[torch.Tensor]): List of tensors we want to hash.
       )");
-  module.def("_dump_nccl_trace", []() {
-    return py::bytes(::c10d::dump_nccl_trace());
-  });
+  module.def(
+      "_dump_nccl_trace",
```
Comment (wconstab) [inline torch/csrc/distributed/c10d/init.cpp:3164]: the names are confusing.  they weren't great before (bc pytorch layer is using 'nccl' name)

but now its extra confusing that the api 'dump_nccl_trace' has an option to exclude the 'traceBuffer' which is further defined as the 'nccl trace buffer' in the docstring.

its pretty unclear to the user what the point of calling `dump_nccl_trace` with includeTraceBuffer=False is.

File: torch/csrc/distributed/c10d/init.cpp [hunk 2]
```diff
@@ -3161,9 +3161,52 @@ such as `dist.all_reduce(tensor, async_op=True)`.
         Arguments:
           tensors(List[torch.Tensor]): List of tensors we want to hash.
       )");
-  module.def("_dump_nccl_trace", []() {
-    return py::bytes(::c10d::dump_nccl_trace());
-  });
+  module.def(
+      "_get_collective_trace",
```
Comment (c-p-i-o) [inline torch/csrc/distributed/c10d/init.cpp:3165]: I made this API with the hopes that there might be cases where we want do debug collectives but not NCCL comm. So this is just a convenient function.
My thought is that collective debugging v/s NCCL comm debugging evolve independently so this is an attempt to start separating these out.

In my mind, the collected data looks as follows:
{
COMMON STUFF (pg_config)
COLLECTIVE_STUFF (entries)
NCCL_COMM_STUFF(a map)
}

File: torch/csrc/distributed/c10d/init.cpp [hunk 3]
```diff
@@ -3161,9 +3161,52 @@ such as `dist.all_reduce(tensor, async_op=True)`.
...
+      R"(
+        Arguments:
+            includeStackTraces(bool, optional): Whether to include stacktraces in the collective work traces. Default is True.
+            onlyActive (bool, optional): Whether to only include active collective work traces. Default is False.
+        Returns:
+            Stringified pickle collective work traces.
+            Default settings return all collectives with stack traces.
+      )");
+  module.def(
+      "_get_nccl_comm_trace",
```
Comment (c-p-i-o) [inline torch/csrc/distributed/c10d/init.cpp:3182]: Similarly, this API helps get at the nccl_comm_trace but not get everything else.

Main idea is that the `get_` API favor retrieval over HTTP

File: torch/csrc/distributed/c10d/init.cpp [hunk 4]
```diff
@@ -3161,9 +3161,52 @@ such as `dist.all_reduce(tensor, async_op=True)`.
...
+      []() { return py::bytes(::c10d::get_nccl_comm_trace()); },
+      R"(
+        Returns:
+            Stringified pickle NCCL comm traces.
+      )");
+  module.def(
+      "_dump_nccl_trace",
```
Comment (c-p-i-o) [inline torch/csrc/distributed/c10d/init.cpp:3189]: This API got a bit confusing - but I left the default behavior as-is and provided some knobs to control data collected.
The main idea is having this for power users who know what they are doing.

**Review by wconstab — APPROVED** [review]
this lgtm.

**Comment by c-p-i-o** [comment]
@pytorchbot merge

I've fixed the internal failing tests in [D58296632](https://www.internalfb.com/diff/D58296632) cc @atalman 
It was a one line fix.

**Comment by c-p-i-o** [comment]
@pytorchbot merge

I've fixed the internal failing tests in [D58296632](https://www.internalfb.com/diff/D58296632) cc @atalman
It was a one line fix.

**Pull request closed without merging**